Make mTLS optional when defining a hook

aaslamin opened the pull request:
**Issue:** https://github.com/aporeto-inc/aporeto/issues/1274

**Context:** when creating a hook, we want to relax the requirement that mTLS must be configured.

This PR makes the destination endpoint TLS configuration optional and adds a custom validator on the model to validate the optional attributes of `clientCertificate` & `clientCertificateKey`. The validator ensures that **if either one of these attributes are provided, then BOTH must be provided** even though the attributes are individually optional. In the event that a validation error is encountered, an elemental error is thrown which provides the user-agent with context on which attributes were provided and missing.

#### TODO: 
- [x] Get early feedback from Chris & Antoine to see if I am headed in the right direction
- [x] Unit tests for custom validator

aaslamin commented:
@t00f @primalmotion Update: I have now added the custom validation here as per your suggestion. 

If it looks A-OK to you, I can start writing the unit tests for the validator 🙏

primalmotion commented on custom_validation_helpers.go:
```diff
@@ -0,0 +1,55 @@
+package gaia
+
```
put that in custom_validations.go near where it used otherwise you’ll have trouble building. we know we need to find a better way to handle this validations but we have cyclic import issue

aaslamin replied:
done

primalmotion commented on custom_validation_helpers.go:
```diff
@@ -0,0 +1,55 @@
+
+	for _, attr := range attrs {
+		switch {
+		case len(attr.value) > 0:
```
for this binary check, please use a if/else. also always check for empty strings using == "". this saves one call, and ensure build breaks if the left operand becomes a map or a slice

aaslamin replied:
> and ensure build breaks if the left operand becomes a map or a slice

Nice, didn't even think of that 😉 

Donezo ✨

primalmotion commented on custom_validation_helpers.go:
```diff
@@ -0,0 +1,55 @@
+			fmt.Sprintf("Attribute(s) [%s] are required because attribute(s) [%s] were provided", strings.Join(missing, ", "), strings.Join(provided, ", ")),
+			"elemental",
+			http.StatusUnprocessableEntity)
+		return err
```
why do you assign it to err? return it directly

aaslamin replied:
Done! 💫

aaslamin commented:
#### Update:

• Moved validation helpers to `custom_validations.go` near where they are used
• Updated docs to be more clear when `clientCertificate` & `clientCertificateKey` need to be provided.

t00f commented on custom_validations.go:
```diff
@@ -414,6 +414,65 @@ func ValidateHTTPMethods(attribute string, methods []string) error {
+// ValidateClientTLSConfig validates the optional 'clientCertificate' & 'clientCertificateKey' attributes of the hook model.
+// The validator ensures that if either one of these attributes are provided, then BOTH should be provided even though the
+// attributes are individually optional.
+func ValidateClientTLSConfig(hp *HookPolicy) error {
```
When reading `ValidateClientTLSConfig` I expect to pass an `TLSConfig` object. As I am passing a hookpolicy, we should call it `ValidateHookPolicy`

aaslamin replied:
Good 👁, updated that now to be more clear! 🙃

aaslamin commented:
#### Update:

• Renamed `ValidateClientTLSConfig` ➡️ `ValidateHookPolicy`

Thanks gents, I will start plumbing some unit tests now for the custom validator.

aaslamin commented:
#### Update

• Added 100% unit test coverage for the new custom validator - `ValidateHookPolicy`

I am ready for another review! 🙏

aaslamin commented on custom_validations_test.go:
```diff
@@ -909,6 +913,96 @@ func TestValidateOptionalNetworkList(t *testing.T) {
+MHcCAQEEIGOXJI/123456789oamOu4tQAIKFdbyvkIJg9GME0mHzoAoGCCqGSM49
+AwEHoUQDQgAE6bM8mP123456789AfmBWtnucfByQXk568lDcKNIQx6yNn+7txbwg
+F9eXFkofGX3UgRtsHe123456789xQ1naSw==
+-----END EC PRIVATE KEY-----`,
```
Nit: I don't actually need to use realistic values here to test the behaviour I am targeting, just thought it would be nice to make it look real 🤷‍♂

primalmotion replied:
Good because soon you will need to :)

primalmotion commented on custom_validations.go:
```diff
@@ -414,6 +414,66 @@ func ValidateHTTPMethods(attribute string, methods []string) error {
+		}
+	}
+
+	if nonEmptyAttrExists && emptyAttrExists {
```
I just realized you don't need these booleans.

```
if len(missing) > 0 &&  len(provided) > 0
```

Always tend to avoid any useless allocation when possible and when it doesn't make the code unreadable.

aaslamin replied:
🦅👁 nice catch!

aaslamin replied:
Donezo

primalmotion commented on custom_validations.go:
```diff
@@ -414,6 +414,66 @@ func ValidateHTTPMethods(attribute string, methods []string) error {
+//
+// which will return an error because Attr2 was empty and Attr1 was set.
+func validateAllOrNoneAttrGrouping(attrs ...attrInfo) error {
+	var missing, provided []string
```
also skip a line between functions prototype and body when more than 1 or 2 lines

aaslamin replied:
Done

primalmotion commented:
/build
